### DISCUSSION TITLE
would io_uring_submit_and_wait()  count cqe from other operations?

### DISCUSSION THREAD

**Discussion by pyhd** [axboe/liburing#1058] [Q&A] [answered]
```
io_uring_prep_recv_multishot(main_ring);
while() {
    io_uring_wait_cqe(main_ring);
    while (io_uring_peek_cqe() == 0) {
        reap & io_uring_prep_sendmsg(send_ring);
        io_uring_cqe_seen(main_ring);
    }
    nr = io_uring_sq_ready(send_ring);
    io_uring_submit_and_wait(send_ring, nr);
    io_uring_cq_advance(send_ring, nr);
}
```

Currently send operation is seperated  from multishot recv ring, because I have to ensure all send operation must be done before next internal recv shot. So my question is: does `io_uring_submit_and_wait` always wait those specifically submitted sqes? Or would it count other cqes if multi-threads `io_uring_submit_and_wait` with the same ring at the same time?

**Comment by axboe** [comment, marked as answer]
There's no way to wait for specific completions, so if you do io_uring_submit_and_wait(ring, nr), then it'll return when 'nr' CQEs are available for you, regardless of where they come from. Generally, if you want to wait for N new receives to come in, you'd probably want to adjust your 'nr' so that it's N+submitted_sends. The sends will generally always complete during submission (unless you're overflowing the socket and it's out of space), hence you should expect the sends you're submitting here to be available in the ring at the time io_uring_submit() (or io_uring_submit_and_wait()) returns control to your application.
